File: atcoder/abc141_e.cpp
// https://atcoder.jp/contests/abc141/tasks/abc141_e
#include<iostream>
#include<algorithm>
//#include<string>
//#include <numeric>
//#include<vector>
//#include<map>
//#include<tuple>
//#include<set>
//#include<queue>
//#include<regex>
//#include <bitset>

// #include <bits/stdc++.h>
#define rep(i,n) for (int i = 0; i < (n); ++i)
using namespace std;
typedef long long ll;

int main() {
	int N;
	cin >> N;
	string S;
	cin >> S;

	int ans = 0;
	rep(i, N - 1) {
		for (int j = i + 1 + ans; j < N - ans; j++) {
			for (int l = ans + 1; l <= min(j - i, N - j); l++) {
				bool chk = true;
				rep(k, l) {
					if (S[i + k] != S[j + k]) {
						chk = false;
						break;
					}
				}
				if (chk) {
					ans = max(ans, l);
				}
				else {
					break;
				}
			}
		}
	}

	cout << ans << endl;
	return 0;
}

File: atcoder/abc141_e.py
# https://atcoder.jp/contests/abc141/tasks/abc141_e

# Pypy
N = int(input())
S = input()
DP = [[0]*5005 for _ in range(5005)]

for i in range(N):
    for j in range(N):
        if S[i] == S[j]:
            DP[N-i][N-j] = DP[N-i+1][N-j+1] + 1
ans = 0
for i in range(N):
    for j in range(N):
        if i >= j:
            continue
        now = min(DP[i][j], j-i)
        ans = max(ans, now)

print(ans)

# # TLE
# N = int(input())
# S = input()

# ans = 0
# for i in range(N-1):
#     for j in range(i+1+ans, N-ans):
#         for l in range(ans+1, min(j-i, N-j)+1):
#             if S[i:i+l] == S[j:j+l]:
#                 ans = max(ans, l)
#             else:
#                 break

# print(ans)
